sql/core/src/main/scala/org/apache/spark/sql/execution/ui/AllExecutionsPage.scala

```diff
@@ -202,100 +385,127 @@ private[ui] abstract class ExecutionTable(
+    val executionUIData: SQLExecutionUIData)
 
-private[ui] class CompletedExecutionTable(
+
+private[ui] class ExecutionDataSource(
+    request: HttpServletRequest,
     parent: SQLTab,
+    executionData: Seq[SQLExecutionUIData],
+    basePath: String,
     currentTime: Long,
-    executionUIDatas: Seq[SQLExecutionUIData])
-  extends ExecutionTable(
-    parent,
-    "completed-execution-table",
-    currentTime,
-    executionUIDatas,
-    showRunningJobs = false,
-    showSucceededJobs = true,
-    showFailedJobs = false) {
+    pageSize: Int,
+    sortColumn: String,
+    desc: Boolean) extends PagedDataSource[ExecutionTableRowData](pageSize) {
 
-  override protected def header: Seq[String] = baseHeader ++ Seq("Job IDs")
-}
+  // Convert ExecutionData to ExecutionTableRowData which contains the final contents to show
+  // in the table so that we can avoid creating duplicate contents during sorting the data
+  private val data = executionData.map(executionRow).sorted(ordering(sortColumn, desc))
 
-private[ui] class FailedExecutionTable(
-    parent: SQLTab,
-    currentTime: Long,
-    executionUIDatas: Seq[SQLExecutionUIData])
-  extends ExecutionTable(
-    parent,
-    "failed-execution-table",
-    currentTime,
-    executionUIDatas,
-    showRunningJobs = false,
-    showSucceededJobs = true,
-    showFailedJobs = true) {
+  private var _slicedJobIds: Set[Int] = _
 
-  override protected def header: Seq[String] =
-    baseHeader ++ Seq("Succeeded Job IDs", "Failed Job IDs")
+  override def dataSize: Int = data.size
+
+  override def sliceData(from: Int, to: Int): Seq[ExecutionTableRowData] = {
+    val r = data.slice(from, to)
+    _slicedJobIds = r.map(_.executionUIData.executionId.toInt).toSet
+    r
+  }
+
+  private def executionRow(executionUIData: SQLExecutionUIData): ExecutionTableRowData = {
+    val submissionTime = executionUIData.submissionTime
+    val duration = executionUIData.completionTime.map(_.getTime())
+      .getOrElse(currentTime) - submissionTime
+
+    new ExecutionTableRowData(
+      submissionTime,
+      duration,
+      executionUIData)
+  }
+
+  /**
+    * Return Ordering according to sortColumn and desc
+    */
+  private def ordering(sortColumn: String, desc: Boolean): Ordering[ExecutionTableRowData] = {
+    val ordering: Ordering[ExecutionTableRowData] = sortColumn match {
+      case "ID" => Ordering.by(_.executionUIData.executionId)
+      case "Description" => Ordering.by(_.executionUIData.description)
+      case "Submitted" => Ordering.by(_.executionUIData.submissionTime)
+      case "Duration" => Ordering.by(_.duration)
+      case "Job IDs" | "Succeeded Job IDs" => Ordering.by(_.executionUIData.jobs.flatMap {
+        case (jobId, jobStatus) =>
+          if (jobStatus == JobExecutionStatus.SUCCEEDED) jobId.toString
+          else ""
+      }.toSeq.toString())
+      case "Running Job IDs" => Ordering.by(_.executionUIData.jobs.flatMap {
+        case (jobId, jobStatus) =>
```

Comment:
Same, here and around here do you really want to filter first instead of flatMap?

Reply:
Thanks. I have modified using filter